atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java

```diff
@@ -80,117 +65,142 @@ private CassandraTimestampBoundStore(CassandraClientPool clientPool) {
-                ColumnOrSuperColumn result;
-                try {
-                    result = client.get(rowName, columnPath, ConsistencyLevel.LOCAL_QUORUM);
-                } catch (NotFoundException e) {
-                    result = null;
-                } catch (Exception e) {
-                    throw Throwables.throwUncheckedException(e);
+        return clientPool.runWithRetry(client -> {
+                    ColumnPath columnPath = new ColumnPath(AtlasDbConstants.TIMESTAMP_TABLE.getQualifiedName());
+                    columnPath.setColumn(ROW_TIMESTAMP_ARRAY);
+
+                    Optional<Column> column = getColumnIfExists(client, ROW_NAME_BYTE_BUFFER, columnPath);
+                    TimestampBoundStoreEntry entryInDb = TimestampBoundStoreEntry.createFromColumn(column);
+                    entryInDb = migrateIfStartingUp(entryInDb);
+                    checkMatchingId(entryInDb);
+                    setCurrentLimit("[GET]", entryInDb);
+                    return currentLimit;
                 }
-                if (result == null) {
-                    DebugLogger.logger.info("[GET] Null result, setting timestamp limit to {}",
-                            CassandraTimestampUtils.INITIAL_VALUE);
-                    cas(client, null, CassandraTimestampUtils.INITIAL_VALUE);
-                    return CassandraTimestampUtils.INITIAL_VALUE;
-                }
-                Column column = result.getColumn();
-                currentLimit = PtBytes.toLong(column.getValue());
-                DebugLogger.logger.info("[GET] Setting cached timestamp limit to {}.", currentLimit);
-                return currentLimit;
-            }
-        });
+        );
     }
 
     @Override
     public synchronized void storeUpperLimit(final long limit) {
         DebugLogger.logger.debug("[PUT] Storing upper limit of {}.", limit);
-        clientPool.runWithRetry(new FunctionCheckedException<Client, Void, RuntimeException>() {
-            @Override
-            public Void apply(Client client) {
-                cas(client, currentLimit, limit);
-                return null;
-            }
+        casWithRetry(TimestampBoundStoreEntry.create(currentLimit, id), TimestampBoundStoreEntry.create(limit, id));
+    }
+
+    private Optional<Column> getColumnIfExists(Client client, ByteBuffer rowName, ColumnPath columnPath) {
+        try {
+            return Optional.of(client.get(rowName, columnPath, ConsistencyLevel.LOCAL_QUORUM).getColumn());
+        } catch (NotFoundException e) {
+            return Optional.empty();
+        } catch (TException e) {
+            throw Throwables.throwUncheckedException(e);
+        }
+    }
+
+    private TimestampBoundStoreEntry migrateIfStartingUp(TimestampBoundStoreEntry entryInDb) {
+        if (startingUp) {
+            DebugLogger.logger.info("[GET] The service is starting up. Attempting to get timestamp bound from the DB "
+                    + "and resetting it with this process's ID!");
```

Comment:
Replace `!` with `.` - Exclaiming this seems too excited, as it's normal process.